core/Enclave/key_factory.cpp

```diff
@@ -0,0 +1,512 @@
+    }
+
+    pem_keypair = (uint8_t*)malloc(key_len + 1);
+    if (pem_keypair == NULL) {
+        goto out;
+    }
+
+    if (BIO_read(bio, pem_keypair, key_len) < 0) {
+        goto out;
+    }
+    pem_keypair[key_len] = '\0';
+
+    ret = ehsm_create_keyblob(pem_keypair, key_len, NULL, 0, (sgx_aes_gcm_data_ex_t*)cmk->keyblob);
+
+    if (ret != SGX_SUCCESS) {
+        goto out;
+    }
+out:
+    EVP_PKEY_CTX_free(pkey_ctx);
+    EVP_PKEY_free(pkey);
+    BIO_free(bio);
+    SAFE_FREE(pem_keypair);
+    return ret;
+}
+
+sgx_status_t ehsm_create_sm4_key(ehsm_keyblob_t *cmk)
+{
+    sgx_status_t ret = SGX_ERROR_UNEXPECTED;
+
+    if (cmk->metadata.keyspec != EH_SM4_CTR &&
+        cmk->metadata.keyspec != EH_SM4_CBC)
+    {
+        return SGX_ERROR_INVALID_PARAMETER;
+    }
+
+    uint32_t keysize = ehsm_get_symmetric_key_size(cmk->metadata.keyspec);
+    if (keysize == 0)
+    {
+        return SGX_ERROR_UNEXPECTED;
+    }
+    uint32_t real_blob_len = ehsm_calc_keyblob_len(keysize);
+
+    if (real_blob_len == UINT32_MAX)
+    {
+        return SGX_ERROR_UNEXPECTED;
+    }
+    if (cmk->keybloblen == 0)
+    {
+        cmk->keybloblen = real_blob_len;
+        return SGX_SUCCESS;
+    }
+
+    uint8_t *tmp = (uint8_t *)malloc(keysize);
+    if (tmp == NULL)
+    {
+        return SGX_ERROR_OUT_OF_MEMORY;
+    }
+    ret = sgx_read_rand(tmp, keysize);
+    if (ret != SGX_SUCCESS)
+    {
+        free(tmp);
+        return ret;
+    }
+    ret = ehsm_create_keyblob(tmp,
+                              keysize,
+                              NULL,
+                              0,
+                              (sgx_aes_gcm_data_ex_t *)cmk->keyblob);
+
+    memset_s(tmp, keysize, 0, keysize);
+
+    free(tmp);
+    return ret;
+}
+sgx_status_t ehsm_generate_datakey_aes(const ehsm_keyblob_t *cmk,
+                                           const ehsm_data_t *aad,
+                                           ehsm_data_t *plaintext,
+                                           ehsm_data_t *ciphertext)
+{
+    sgx_status_t ret = SGX_ERROR_UNEXPECTED;
+
+    uint8_t *temp_datakey =NULL;
+    temp_datakey = (uint8_t *)malloc(plaintext->datalen);
+    if (temp_datakey == NULL)
+    {
+        free(temp_datakey);
+        return SGX_ERROR_OUT_OF_MEMORY;
+    }
+    if(RAND_bytes(temp_datakey, plaintext->datalen) != 1)
+    {
+        free(temp_datakey);
+        return SGX_ERROR_OUT_OF_MEMORY;
+    }
+
+    memcpy_s(plaintext->data, plaintext->datalen, temp_datakey, plaintext->datalen);
+
+    ret = ehsm_aes_gcm_encrypt(aad,
+                               cmk,
+                               plaintext,
+                               ciphertext);
+
+   free(temp_datakey);
+
+    return ret;
+}
+
+sgx_status_t ehsm_generate_datakey_sm4(const ehsm_keyblob_t *cmk,
+                                       ehsm_data_t *plaintext,
+                                       ehsm_data_t *ciphertext)
+{
+    sgx_status_t ret = SGX_ERROR_UNEXPECTED;
+
```

Comment:
is the cmk buffer length checked outside enclave?

Reply:
will be added